flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPoolBridge.java

```diff
@@ -0,0 +1,680 @@
+		declareResourceRequirements(declarativeSlotPool.getResourceRequirements());
+	}
+
+	@Override
+	public void disconnectResourceManager() {
+		assertRunningInMainThread();
+		this.declareResourceRequirementServiceConnectionManager.disconnect();
+	}
+
+	@Override
+	public boolean registerTaskManager(ResourceID resourceID) {
+		assertRunningInMainThread();
+
+		LOG.debug("Register new TaskExecutor {}.", resourceID);
+		return registeredTaskManagers.add(resourceID);
+	}
+
+	@Override
+	public boolean releaseTaskManager(ResourceID resourceId, Exception cause) {
+		assertRunningInMainThread();
+
+		if (registeredTaskManagers.remove(resourceId)) {
+			internalReleaseTaskManager(resourceId, cause);
+			return true;
+		} else {
+			return false;
+		}
+	}
+
+	private void releaseAllTaskManagers(FlinkException cause) {
+		for (ResourceID registeredTaskManager : registeredTaskManagers) {
+			internalReleaseTaskManager(registeredTaskManager, cause);
+		}
+
+		registeredTaskManagers.clear();
+	}
+
+	private void internalReleaseTaskManager(ResourceID resourceId, Exception cause) {
+		ResourceCounter previouslyFulfilledRequirement = declarativeSlotPool.releaseSlots(resourceId, cause);
+		declarativeSlotPool.decreaseResourceRequirementsBy(previouslyFulfilledRequirement);
+	}
+
+	@Override
+	public Collection<SlotOffer> offerSlots(TaskManagerLocation taskManagerLocation, TaskManagerGateway taskManagerGateway, Collection<SlotOffer> offers) {
+		assertRunningInMainThread();
+		Preconditions.checkNotNull(taskManagerGateway);
+		Preconditions.checkNotNull(offers);
+
+		if (!registeredTaskManagers.contains(taskManagerLocation.getResourceID())) {
+			return Collections.emptyList();
+		}
+
+		return declarativeSlotPool.offerSlots(offers, taskManagerLocation, taskManagerGateway, clock.relativeTimeMillis());
+	}
+
+	@VisibleForTesting
+	void newSlotsAreAvailable(Collection<? extends PhysicalSlot> newSlots) {
+		// Avoid notifying new slots multiple times due to SchedulerImpl allocating and releasing slots
+		// in order to find the best shared slot
+		final Collection<PhysicalSlot> slotsToProcess = new ArrayList<>();
+		for (PhysicalSlot newSlot : newSlots) {
+			if (newSlotsSet.add(newSlot.getAllocationId())) {
+				slotsToProcess.add(newSlot);
+			}
+		}
+
+		final Collection<PendingRequestSlotMatching> matchingsToFulfill = new ArrayList<>();
+
+		for (PhysicalSlot newSlot : slotsToProcess) {
+			final Optional<PendingRequest> matchingPendingRequest = findMatchingPendingRequest(newSlot);
+
+			matchingPendingRequest.ifPresent(pendingRequest -> {
+				Preconditions.checkNotNull(pendingRequests.remove(pendingRequest.getSlotRequestId()), "Cannot fulfill a non existing pending slot request.");
+				reserveFreeSlot(pendingRequest.getSlotRequestId(), newSlot.getAllocationId(), pendingRequest.resourceProfile);
+
+				matchingsToFulfill.add(PendingRequestSlotMatching.createFor(pendingRequest, newSlot));
+			});
+
+			newSlotsSet.remove(newSlot.getAllocationId());
```

Comment:
In the current state I also cannot see a point in the `newSlotsSet`. I suppose it only had any effect if the request completion triggers another free/reserve cycle, i.e., it triggers a recursion.
With the completion moved into a separate loop the reservation of slots is now locked in before any recursion can happen, which seems...good?

What I'm unsure of are the implications. I think we're fine from a _correctness_ point of view, but does this mean the scheduler now has less options when it comes to how shared slots are allocated?

TBH I don't fully understand the issue you fixed in https://github.com/tillrohrmann/flink/commit/2c02f54a35cbcd7f5f4148062f5cd656b3a41810. It reads to me like the scheduler is cycling through different allocation setups, but if it frees the slot in between how can a vertex be assigned twice? Was the slot pool leaking state across recursions, or is the scheduler doing the book-keeping incorrectly?

Unfortunately I don't quite remember when and why I moved the completion out of the loop; but not having this recursion should at least make it easier to reason about...

Reply:
I think this might have tried to solve a similar problem as tillrohrmann@8c6f00ef7d26bc47459a0b369976de8581e5e8b3 but I also cannot recall what the problematic scenario was. Looking at the code, I think we can remove it.

Reply:
I think I could reproduce the problem using the `TimeStampITCase` with the commit before tillrohrmann@2c02f54. The problem was that we were reusing the same `MultiTaskSlot` for the same `JobVertexID` twice. I think the problem is the following: When scheduling a subtask `a`, we allocate a `MultiTaskSlot` in `SchedulerImpl.allocateMultiTaskSlot` by picking a `MultiTaskSlot` which does not contain yet the `JobVertexID`. Once we have it, we try to allocate an available slot to see whether it has a better locality. If this is not the case, then we release the slot. This will trigger the `newSlotsAreAvailable` method again. If now the last input for a subtask `b` of the same `JobVertexID` is pending, the released slot could be used to fulfill the request. This would then trigger the scheduling of `b` which could take the same `MultiTaskSlot` which `a` had picked but not yet reserved. The problem is that due to the direct executions of the future, we can have an interleaved execution where multiple calls to `SchedulerImpl.allocateMultiTaskSlot` can happen nestedly w/o being properly terminated.

Reply:
Long story short, this should no longer a problem with first assigning the slots to pending requests and then fulfilling them.